rclpy/rclpy/node.py

```
@@ -168,57 +169,59 @@ def __init__(
+        self._pre_set_parameters_callbacks: List[Callable[[List[Parameter[Any]]],
+                                                          List[Parameter[Any]]]] = []
         self._on_set_parameters_callbacks: \
-            List[Callable[[List[Parameter]], SetParametersResult]] = []
-        self._post_set_parameters_callbacks: List[Callable[[List[Parameter]], None]] = []
+            List[Callable[[List[Parameter[Any]]], SetParametersResult]] = []
+        self._post_set_parameters_callbacks: List[Callable[[List[Parameter[Any]]], None]] = []
         self._rate_group = ReentrantCallbackGroup()
         self._allow_undeclared_parameters = allow_undeclared_parameters
-        self._parameter_overrides: Dict[str, Parameter] = {}
+        self._parameter_overrides: Dict[str, Parameter[Any]] = {}
         self._descriptors: Dict[str, ParameterDescriptor] = {}
 
         namespace = namespace or ''
         if not self._context.ok():
             raise NotInitializedException('cannot create node')
-        with self._context.handle:
-            try:
-                self.__node = _rclpy.Node(
-                    node_name,
-                    namespace,
-                    self._context.handle,
-                    cli_args,
-                    use_global_arguments,
-                    enable_rosout
-                )
-            except ValueError:
-                # these will raise more specific errors if the name or namespace is bad
-                validate_node_name(node_name)
-                # emulate what rcl_node_init() does to accept '' and relative namespaces
-                if not namespace:
-                    namespace = '/'
-                if not namespace.startswith('/'):
-                    namespace = '/' + namespace
-                validate_namespace(namespace)
-                # Should not get to this point
-                raise RuntimeError('rclpy_create_node failed for unknown reason')
+        if self._context.handle:
```

This changes the behavior if that is NULL? before it catches `ValueError` but after it passes this if statement and proceed to the following code?

---

Do you think it would be better to raise an Exception if it is `None` rather than attempt to carry on with the rest of the `__init__` function?

---

Ok I switched it to raising an exception if it does not exist.

---

Another solution to potentially avoid all these `context.handle` None checks would be to move the Context's `init` function directly into the `__init__` function. That way the context handle would always exist. However, there might be some reason I am unaware of for this delayed construction.